{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import ipywidgets as ipw\n",
    "import numpy as np\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#simple helper functions\n",
    "def gmean(data, axis=0):\n",
    "    return np.exp(np.mean(np.log(data), axis=axis))\n",
    "\n",
    "def combine_ranks(data, hpo_terms):\n",
    "    # get geometric mean of ranks for given terms\n",
    "    hpo_terms = [x for x in hpo_terms if x in data.columns] # remove terms not in dataset    \n",
    "    return pd.Series(gmean(data.loc[:, hpo_terms], axis=1), index=data.index)\n",
    "\n",
    "def reduce_terms(data, hpo_tree, hpo_terms):\n",
    "    #given a list of hpo terms, get the closest ancestor in dataset columns\n",
    "    term_list = set()\n",
    "    for term in hpo_terms:\n",
    "        #get the list of terms that actually are in dataset\n",
    "        path = [x for x in nx.shortest_path(hpo_tree, term, 'HP:0000118') if x in data.columns]\n",
    "        if len(path) > 0:\n",
    "            term_list.add(path[0])\n",
    "    return term_list\n",
    "\n",
    "def build_table(df):\n",
    "    spool = \"\"\"\n",
    "    <table width='100%'>\n",
    "    <tr>\n",
    "    <th>Entrez ID</th>\n",
    "    <th>Gene Symbol</th>\n",
    "    <th>Name</th>\n",
    "    <th>Score</th>\n",
    "    </tr>\"\"\"\n",
    "    for gene in df.index:\n",
    "        g_name = gene_annotation.node[gene]['name']\n",
    "        g_symbol = list(gene_annotation.neighbors(gene))[0]\n",
    "        spool += \"<tr><td>%s</td><td>%s</td><td>%s</td><td>%.4f</td></tr>\" % (gene, g_symbol, g_name, df.loc[gene])\n",
    "    spool += \"</table>\"\n",
    "    return spool\n",
    "\n",
    "def get_valid_genes(gene_string):\n",
    "    # first retain only genes that are in our annotation\n",
    "    genes = gene_string.split('\\n')\n",
    "    final_genes = set()\n",
    "    genes_in = [x for x in genes if x in gene_annotation.nodes()]\n",
    "    genes_out = [x for x in genes if not x in gene_annotation.nodes()]\n",
    "    for g in genes_in:\n",
    "        if gene_annotation.node[g]['type'] == 'entrez':\n",
    "            final_genes.add(g)\n",
    "        else:\n",
    "            #a gene symbol is given\n",
    "            [final_genes.add(x) for x in gene_annotation.neighbors(g)]\n",
    "    return final_genes, genes_out\n",
    "\n",
    "def add_warning_excl(genes_out):\n",
    "    spool = '<hr><div class=\"excluded\"><b>Warning!</b> The following items were excluded from the analysis:<br>'\n",
    "    for g in genes_out:\n",
    "        spool += '%s<br>' % g\n",
    "    spool += \"</div>\"\n",
    "    return spool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data\n",
    "#github does not allow files larger than 100 MB, so I splitted our model\n",
    "model_ranks = pd.read_pickle(\"data/global_ind_pheno_tumor_munge.best_pred.00.pickle\")\n",
    "for chunk in range(1, 9):\n",
    "    model_ranks = pd.concat([model_ranks, pd.read_pickle(\"data/global_ind_pheno_tumor_munge.best_pred.%02d.pickle\" % chunk)])\n",
    "    \n",
    "#scored terms with some annotation\n",
    "hp_annot = pd.read_pickle(\"data/global_ind_pheno_tumor_munge.best.pickle\")\n",
    "#combined ranks for OMIM diseases\n",
    "disease_net = nx.gpickle.read_gpickle(\"data/disease_annot.gpickle\")\n",
    "#the HPO obo tree\n",
    "hpo_net = nx.gpickle.read_gpickle(\"data/hp.180127.obo.gpickle\")\n",
    "#gene annotation to have gene symbol and/or entrez ID\n",
    "gene_annotation = nx.gpickle.read_gpickle(\"data/gene_annotation.gpickle\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpo_ids = [x for x in hpo_net.nodes() if nx.has_path(hpo_net, x, 'HP:0000118')]\n",
    "hpo_ids.sort()\n",
    "hpo_ids = hpo_ids[1:]\n",
    "hpo_names = [hpo_net.node[x]['name'] for x in hpo_ids]\n",
    "hpo_terms = [\"%s: %s\" % (hpo_ids[x], hpo_names[x]) for x in range(len(hpo_ids))]\n",
    "disease_ids = [x for x in disease_net.nodes() if disease_net.node[x]['type'] == 'disease']\n",
    "disease_ids.sort()\n",
    "disease_names = [disease_net.node[x]['name'] for x in disease_ids]\n",
    "diseases = [\"%s: %s\" % (disease_ids[x], disease_names[x]) for x in range(len(disease_ids))]\n",
    "selected_terms = set()\n",
    "\n",
    "#define the widget for HPO search\n",
    "#the elements\n",
    "gene_list = ipw.Textarea(\n",
    "    placeholder='Enter your gene list',\n",
    "    description='Genes:',\n",
    "    disabled=False\n",
    ")\n",
    "\n",
    "search_hp_widget = ipw.Text(placeholder='Start typing phenotypes...') #to search the phenotypes\n",
    "search_dis_widget = ipw.Text(placeholder='Start typing description...') #to search the phenotypes\n",
    "options_widget = ipw.SelectMultiple(options=hpo_terms) #this lists all terms to be selected\n",
    "disease_widget = ipw.Select(options=diseases)\n",
    "add_hp_button = ipw.Button(description='Add term')\n",
    "remove_button = ipw.Button(description='Remove term')\n",
    "add_dis_button = ipw.Button(description='Add disease')\n",
    "selected_widget = ipw.SelectMultiple()\n",
    "submit_button = ipw.Button(description='Submit')\n",
    "reset_button = ipw.Button(description='Reset')\n",
    "out = ipw.Output()\n",
    "results = ipw.HTML()\n",
    "\n",
    "#stitch together\n",
    "hp_search = ipw.VBox([search_hp_widget, options_widget, add_hp_button])\n",
    "dis_search = ipw.VBox([search_dis_widget, disease_widget, add_dis_button])\n",
    "chosen_area = ipw.VBox([selected_widget, remove_button])\n",
    "bottom_area = ipw.HBox([reset_button, submit_button])\n",
    "selection_area = ipw.HBox([hp_search, ipw.HTML(value=\"<b> OR </b>\"), dis_search])\n",
    "multi_select = ipw.VBox([selection_area, chosen_area])\n",
    "\n",
    "#define actions\n",
    "def on_hp_search(change):\n",
    "    search_input = change['new']\n",
    "    if search_input == '':\n",
    "        # Reset search field\n",
    "        new_options = hpo_terms\n",
    "    else:\n",
    "        # Filter by search field \n",
    "        new_options = [x for x in hpo_terms if search_input.lower() in x.lower()]\n",
    "    options_widget.options = new_options\n",
    "\n",
    "def on_dis_search(change):\n",
    "    search_input = change['new']\n",
    "    if search_input == '':\n",
    "        # Reset search field\n",
    "        new_options = diseases\n",
    "    else:\n",
    "        # Filter by search field \n",
    "        new_options = [x for x in diseases if search_input.lower() in x.lower()]\n",
    "    disease_widget.options = new_options\n",
    "\n",
    "    \n",
    "def on_add_hp(b):\n",
    "    new_options = set([x for x in selected_widget.options] + [x for x in options_widget.value])\n",
    "    new_options = list(new_options)\n",
    "    new_options.sort()\n",
    "    selected_widget.options = new_options\n",
    "\n",
    "def on_add_dis(b):\n",
    "    dis_id = disease_widget.value[:11]\n",
    "    hp_map = [x for x in disease_net.neighbors(dis_id) if nx.has_path(hpo_net, x, 'HP:0000118')]\n",
    "    hp_names = [hpo_net.node[x]['name'] for x in hp_map]\n",
    "    hp_terms = [\"%s: %s\" % (hp_map[x], hp_names[x]) for x in range(len(hp_map))]\n",
    "    new_options = set([x for x in selected_widget.options] + hp_terms)\n",
    "    new_options = list(new_options)\n",
    "    new_options.sort()\n",
    "    selected_widget.options = new_options\n",
    "\n",
    "    \n",
    "    \n",
    "def on_remove(b):\n",
    "    v = [x for x in selected_widget.options if not x in selected_widget.value]\n",
    "    selected_widget.options = v\n",
    "\n",
    "def on_submit(b):\n",
    "    input_genes = []\n",
    "    input_terms = []\n",
    "    excluded_genes = []\n",
    "    if not gene_list.value:\n",
    "        output = \"<b>You should provide some genes</b>\"\n",
    "    else:\n",
    "        input_genes, excluded_genes = get_valid_genes(gene_list.value)\n",
    "    if not selected_widget.options:\n",
    "        output = \"<b>You should provide some HPO terms</b>\"\n",
    "    input_terms = [x[:10] for x in selected_widget.options]\n",
    "    if not input_genes:\n",
    "        output = \"<b>None of the genes provided is valid</b>\"\n",
    "    elif input_terms:\n",
    "        #we have genes and terms\n",
    "        subset_ranks = model_ranks.loc[input_genes] #subset data to perform less calculation\n",
    "        ranked_genes = combine_ranks(subset_ranks, reduce_terms(subset_ranks, hpo_net, input_terms)).sort_values()\n",
    "        output = build_table(ranked_genes)\n",
    "    if excluded_genes:\n",
    "        output += add_warning_excl(excluded_genes)\n",
    "    results.value = output\n",
    "\n",
    "\n",
    "def on_reset(b):\n",
    "    search_hp_widget.value = ''\n",
    "    search_dis_widget.value = ''\n",
    "    gene_list.value = ''\n",
    "    selected_widget.options = ()\n",
    "    results.value = ''\n",
    "\n",
    "#link actions\n",
    "add_hp_button.on_click(on_add_hp)\n",
    "add_dis_button.on_click(on_add_dis)                      \n",
    "remove_button.on_click(on_remove)    \n",
    "submit_button.on_click(on_submit)\n",
    "reset_button.on_click(on_reset)\n",
    "search_hp_widget.observe(on_hp_search, names='value')\n",
    "search_dis_widget.observe(on_dis_search, names='value')\n",
    "#display\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DISCAN notebook\n",
    "This is a simple implementation of [DISCAN](https://doi.org/10.1101/120121) using Jupyter interactive notebook. \n",
    "\n",
    "This web application is intended to score a gene list against a given set of phenotypes encoded in the [Human Phenotype Ontology](http://human-phenotype-ontology.github.io). Phenotypes can be added by selecting a specifc disease from the disease menu.\n",
    "\n",
    "---\n",
    "**Warning**\n",
    "\n",
    "*If you are using this notebook on [binder](http://mybinder.org), you may expect poor performances*\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1) Enter a list of genes here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2) Choose some phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(multi_select)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3) Submit your query and check results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(bottom_area, results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "<small>\n",
    "Davide Cittaro<br>\n",
    "[Center for Translational Genomics and Bioinformatics, Milano](http://www.hsr.it/research/organization/divisions-centers/center-for-translational-genomics-and-bioinformatics/)\n",
    "</small>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
